File: medium/0002_add_two_numbers/solution.py
from typing import Optional

from utils.python.nodes import ListNode


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = ListNode(-1)
        curr = dummy

        carry = 0

        while l1 or l2 or carry:
            if l1 and l2:
                new_val = l1.val + l2.val
                l1 = l1.next
                l2 = l2.next
            elif l1:
                new_val = l1.val
                l1 = l1.next
            elif l2:
                new_val = l2.val
                l2 = l2.next
            else:
                new_val = 0

            new_val += carry
            carry = new_val > 9

            curr.next = ListNode(new_val % 10)
            curr = curr.next

        return dummy.next

File: medium/0002_add_two_numbers/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  utils::ListNode *addTwoNumbers(utils::ListNode *l1, utils::ListNode *l2) {
    utils::ListNode dummy(-1);
    utils::ListNode *curr = &dummy;

    int carry = 0;
    int val1 = 0, val2 = 0, newVal;

    while (l1 != nullptr || l2 != nullptr || carry) {
      val1 = l1 != nullptr ? l1->val : 0;
      val2 = l2 != nullptr ? l2->val : 0;

      l1 = l1 != nullptr ? l1->next : nullptr;
      l2 = l2 != nullptr ? l2->next : nullptr;

      newVal = val1 + val2 + carry;
      carry = newVal / 10;

      curr->next = new utils::ListNode(newVal % 10);
      curr = curr->next;
    }

    return dummy.next;
  }
};
